plugin/template/template.go

```diff
@@ -0,0 +1,143 @@
+package template
+
+import (
+	"bytes"
+	"context"
+	"regexp"
+	"strconv"
+	gotmpl "text/template"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// Handler is a plugin handler that takes a query and templates a response.
+type Handler struct {
+	Next      plugin.Handler
+	Templates []template
+}
+
+type template struct {
+	rcode      int
+	class      uint16
+	qtype      uint16
+	regex      []*regexp.Regexp
+	answer     []*gotmpl.Template
+	additional []*gotmpl.Template
+}
+
+type templateData struct {
+	Name     string
+	Regex    string
+	Match    []string
+	Group    map[string]string
+	Class    string
+	Type     string
+	Message  *dns.Msg
+	Question *dns.Question
+}
+
+// ServeDNS implements the plugin.Handler interface.
+func (h Handler) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	for _, template := range h.Templates {
+		data, match := template.match(r)
+		if !match {
+			continue
+		}
+
+		TemplateMatchesCount.WithLabelValues(data.Regex).Inc()
+
+		if template.rcode == dns.RcodeServerFailure {
+			return template.rcode, nil
+		}
+
+		msg := new(dns.Msg)
+		state := request.Request{W: w, Req: r}
+		msg.SetReply(r)
+		msg.Authoritative, msg.RecursionAvailable, msg.Compress = true, true, true
+		msg.Rcode = template.rcode
+
+		for _, answer := range template.answer {
+			rr, err := executeRRTemplate("answer", answer, data)
+			if err != nil {
+				return dns.RcodeServerFailure, err
+			}
+			msg.Answer = append(msg.Answer, rr)
+		}
+		for _, additional := range template.additional {
+			rr, err := executeRRTemplate("additional", additional, data)
+			if err != nil {
+				return dns.RcodeServerFailure, err
+			}
+			msg.Extra = append(msg.Extra, rr)
+		}
+
+		state.SizeAndDo(msg)
+		w.WriteMsg(msg)
+		return template.rcode, nil
+	}
+	return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+}
+
+// Name implements the plugin.Handler interface.
+func (h Handler) Name() string {
+	return "template"
+}
+
+func executeRRTemplate(section string, template *gotmpl.Template, data templateData) (dns.RR, error) {
+	buffer := &bytes.Buffer{}
+	err := template.Execute(buffer, data)
+	if err != nil {
+		TemplateFailureCount.WithLabelValues(data.Regex, section, template.Tree.Root.String()).Inc()
```

Comment:
Slightly nice to let the caller increment the failure counter? I.e similar to not log errors in a function, but just returning the error?

Reply:
This function combines 2 steps and their error handling.

Unrolling the error in the caller is close to inlining this function in the first place. Happy to do that if it makes sense.

Reply:
nah, leave it for this then.